server/src/main/java/io/crate/replication/logical/MetadataTracker.java

```diff
@@ -376,19 +376,29 @@ static ClusterState updateIndexMetadata(String subscriptionName,
      * These tables will be reported
      * and will be visible in pg_subscriptions_rel table.
      */
-    private static List<RelationName> getExistingLocallyTables(Subscription subscription,
+    private static Set<RelationName> getExistingLocallyTables(Subscription subscription,
                                                                ClusterState subscriberClusterState,
-                                                               Collection<RelationName> updatedRelationsList) {
-        // Table existence check is done on a subscription creation but we need such check on
-        // every remote state update to cover case when several clusters' subscription chain forms a loop,
-        // no tables at the beginning and one of the cluster subscribed to ALL TABLES.
-        // In such case the cluster can transitively subscribe to its own table.
+                                                               Response publisherStateResponse) {
+        // Table existence check is done on a subscription creation but
+        // there are cases when attempt to subscribe to existing table happens later during the replication.
         var metadata = subscriberClusterState.metadata();
         Set<RelationName> currentlyReplicatedTables = subscription.relations().keySet();
-        return updatedRelationsList.stream()
+
+        Set<RelationName> existingRelations = publisherStateResponse.concreteIndices().stream()
+            .filter(index -> metadata.hasIndex(index))
+            .map(index -> RelationName.fromIndexName(index))
+            .filter(relationName -> currentlyReplicatedTables.contains(relationName) == false)
+            .collect(Collectors.toSet());
+
+        existingRelations.addAll(
+            publisherStateResponse.concreteTemplates().stream()
+            .filter(t -> metadata.templates().containsKey(t))
+            .map(t -> PartitionName.fromIndexOrTemplate(t).relationName())
             .filter(r -> currentlyReplicatedTables.contains(r) == false)
-            .filter(r -> metadata.hasIndex(r.indexNameOrAlias()))
-            .collect(Collectors.toList());
+            .collect(Collectors.toSet())
```

Comment:
I think adding each item one by one may be faster in this case as afaik `addAll` on a (Hash)Set is not optimized in some way and this here requires intermediate sets. But probably a micro optimization, so  fine to keep as is as well.